pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -711,21 +717,23 @@ func newUnschedulablePodsMap(metricRecorder metrics.MetricRecorder) *Unschedulab
 	}
 }
 
-// nominatedPodMap is a structure that stores pods nominated to run on nodes.
+// NominatedPodMap is a structure that stores pods nominated to run on nodes.
 // It exists because nominatedNodeName of pod objects stored in the structure
 // may be different than what scheduler has here. We should be able to find pods
 // by their UID and update/delete them.
-type nominatedPodMap struct {
+type NominatedPodMap struct {
 	// nominatedPods is a map keyed by a node name and the value is a list of
 	// pods which are nominated to run on the node. These are pods which can be in
 	// the activeQ or unschedulableQ.
 	nominatedPods map[string][]*v1.Pod
 	// nominatedPodToNode is map keyed by a Pod UID to the node name where it is
 	// nominated.
 	nominatedPodToNode map[ktypes.UID]string
+
+	lock sync.RWMutex
```

Comment:
It doesn't seem we need the lock. Accesses from the queue are already synced using the queue's lock. Accesses from generic_scheduler.go are not happening in parallel.

Reply:
ugh, ignore that, we need the lock to sync between accesses happening by the queue (triggered by events) and generic_scheduler.go.

I suggest we embed the lock and just use npm.Lock/Unlock.

Reply:
I suppose you mean embed anonymous `sync.RWMutex` and so we can use npm.Lock/Unlock?

Reply:
>I suppose you mean embed anonymous sync.RWMutex and so we can use npm.Lock/Unlock?

Yes. 

Thinking about this a bit more, I worry that we now have a second mutex, I am concerned about potential deadlocks. The original code was simpler.

How about we keep everything as is, and just use the queue itself as the backing implementation of ```PodNominator``` interface?

Reply:
I was thinking of that at the beginning. One primary disadvantage is that the lock is too wide - for the operation that only concern Nominating, we're locking the whole schedulingQueue, although it's the original way, but IMO is not fine-grained.

Another benefit of separate PodNominator from schedulingQueue is we can just inject a NominatedMap to the places that just need a PodNominator, like the plugins, tests, etc.